src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,444 @@
+{
+    CHIP_ERROR err   = CHIP_NO_ERROR;
+    uint8_t rangeCtl = 0, ptcByte = 0;
+
+    // BufBound has to be initialized with aBuffer so VerifyOrExit can't be used here.
+    if (aBuffer == NULL)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    BufBound bbuf(aBuffer->Start(), aBuffer->AvailableDataLength());
+
+    ptcByte |= mSupportedVersions & VERSION_MASK;
+    if (mSupportsSenderDrive)
+        ptcByte |= SENDER_DRIVE_MASK;
+    if (mSupportsReceiverDrive)
+        ptcByte |= RECEIVER_DRIVE_MASK;
+    if (mSupportsAsync)
+        ptcByte |= ASYNC_MASK;
+
+    bbuf.Put(ptcByte);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    if (mDefLen)
+        rangeCtl |= DEFLEN_MASK;
+    if (mStartOffset)
+        rangeCtl |= START_OFFSET_MASK;
+    if (mWideRange)
+        rangeCtl |= WIDERANGE_MASK;
+
+    bbuf.Put(rangeCtl);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    bbuf.PutLE16(mMaxBlockSize);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    if (mStartOffset > 0)
+    {
+        if (mWideRange)
+        {
+            bbuf.PutLE64(mStartOffset);
+            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        }
+        else
+        {
+            bbuf.PutLE32((uint32_t) mStartOffset);
+            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        }
+    }
+
+    if (mDefLen)
+    {
+        if (mWideRange)
+        {
+            bbuf.PutLE64(mMaxLength);
+            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        }
+        else
+        {
+            bbuf.PutLE32((uint32_t) mMaxLength);
+            VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        }
+    }
+
+    VerifyOrExit(mFileDesignator && mFileDesLength > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
+    bbuf.PutLE16(mFileDesLength);
+    bbuf.Put(mFileDesignator, (size_t) mFileDesLength);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Metadata is optional
+    if (mMetadata && mMetadataLength > 0)
+    {
+        bbuf.Put(mMetadata, (size_t) mMetadataLength);
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+    }
+
+    aBuffer->SetDataLength(static_cast<uint16_t>(bbuf.Written()));
+
+exit:
+    return err;
+}
+
+CHIP_ERROR TransferInit::Parse(System::PacketBuffer * aBuffer, TransferInit & aParsedMessage)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint8_t ptcByte;
+    uint8_t rangeCtl;
+    bool hasStartOffset     = false;
+    uint32_t tmpUint32Value = 0;
+    uint8_t * bufStart;
+
+    // BufBound has to be initialized with aBuffer so VerifyOrExit can't be used here.
+    if (aBuffer == NULL)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    bufStart = aBuffer->Start();
+    Reader bufReader(bufStart, aBuffer->DataLength());
+
+    SuccessOrExit(bufReader.Read8(&ptcByte).StatusCode());
```

Comment:
You can probably combine these reads into a single sequence of reads, since the later ones don't depend on the results of the former.  So:
```
  SuccessOrExit bufReader.Read8(&ptcByte).Read8(&rangeCtl).Read16(&aParsedMessage.mMaxBlockSize).StatusCode());
```
and then examine those values, etc.  The later reads _do_ depend on these values, so can't be commoned up like that.

Reply:
Good point, this will reduce lines of code